### PR TITLE
Local function test categorization

### PR DESCRIPTION
**Pull request by jaredpar**

### CODE REVIEW

**Comment by jaredpar** [comment]
In a recent stand up I mentioned using xunit traits as a method to begin categorizing our tests.  Essentially giving us the ability to conduct queries on our test bed like:

- How many local function tests do we have? 
- What is our coverage of local function + dynamic? 

This is an example of how that system would work and I wanted to get the compiler teams feedback on it.  

I'm happy with the first item: marking tests intended for a specific feature.  It's important for us to be able to examine the coverage we've added at any given time for a particular feature.  Traits are a good way of doing this as they are supported by the underlying tools we use. 

I'm less certain about doing the intersection: the places where I marked specific local function tests as also working with say dynamic, async, etc ...  It's valuable but I'm worried it could also get tedious.  Curious about others thoughts here.

CC @dotnet/roslyn-compiler

File: src/Test/Utilities/Desktop/CompilerTraitAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Xunit;
+using Xunit.Sdk;
+
+namespace Microsoft.CodeAnalysis.Test.Utilities
+{
+    [TraitDiscoverer("Microsoft.CodeAnalysis.Test.Utilities.CompilerTraitDiscoverer", assemblyName: "Roslyn.Test.Utilities.Desktop")]
```
Comment (gafter) [inline src/Test/Utilities/Desktop/CompilerTraitAttribute.cs]: Where is the documentation for `TraitDiscoverer`?

File: src/Test/Utilities/Desktop/CompilerTraitAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
+    public sealed class CompilerTraitAttribute : Attribute, ITraitAttribute
+    {
+        public CompilerFeature Feature { get; }
+
+        public CompilerTraitAttribute(CompilerFeature feature)
```
Comment (gafter) [inline src/Test/Utilities/Desktop/CompilerTraitAttribute.cs]: Please make this params so we can tag multiple "traits" in the same attribute.

File: src/Test/Utilities/Desktop/CompilerTraitAttribute.cs (hunk 1 above)
Comment (jaredpar) [inline src/Test/Utilities/Desktop/CompilerTraitAttribute.cs]: There are a couple of places to read about:

- http://mac-blog.org.ua/xunit-category-trait/
- https://dennymichael.net/2015/01/23/how-to-extendmigrate-traitattribute-to-xunit-v2-and-why-has-become-sealed/

The purpose of the discoverer though is just to map the trait to a name / value pair.  This is how xunit fundamentally understands traits.  Their core TraitAttribute has the form `[Trait(name, value)]` and the discoverer just maps back the custom attribute to their understanding.

I have verified this implementation works correctly with the standard xunit libraries.

File: src/Test/Utilities/Desktop/CompilerTraitAttribute.cs (hunk 2 above)
Comment (jaredpar) [inline src/Test/Utilities/Desktop/CompilerTraitAttribute.cs]: Had not thought of that idea.  I'm fairly certain I can get that to work but I need to check.

**Comment by gafter** [comment]
I think this is a fine thing to do, but I would not suggest we push for any requirement of categorizing all of the existing tests at any particular time, at least not soon. This would be a nice thing to open it to community contributions. I think it would make sense to tag all new tests as they are added, and tag older tests on an opportunistic basis.

**Comment by jaredpar** [comment]
> I think it would make sense to tag all new tests as they are added, and tag older tests on an opportunistic basis.

Completely agree.

File: src/Test/Utilities/Desktop/CompilerTraitAttribute.cs (hunk 2 above)
Comment (jaredpar) [inline src/Test/Utilities/Desktop/CompilerTraitAttribute.cs]: Yep that works great.  Just uploaded again with a `params` version.

File: src/Test/Utilities/Portable/CompilerTraitAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (gafter) [inline src/Test/Utilities/Portable/CompilerTraitAttribute.cs]: This appears identical to `src/Test/Utilities/Desktop/CompilerTraitAttribute.cs`. Do we need both?

Comment (jaredpar) [inline src/Test/Utilities/Portable/CompilerTraitAttribute.cs]: Unfortunately we need both.  The `TraitDiscoverer` attribute specifies the assembly where the actual discoverer object lives.  The name is different between our portable and desktop version.

Comment (gafter) [inline src/Test/Utilities/Portable/CompilerTraitAttribute.cs]: Sure, but can't we refer to the same source file in both projects, like we do for some other sources?

Comment (jaredpar) [inline src/Test/Utilities/Portable/CompilerTraitAttribute.cs]: No.  The TraitDiscoverer attribute does not allow multiple uses and it has to have differet values for the assemblyName parameter.  It's unfortunate.  I'd raelly like to not do this.

**Comment by gafter** [comment]
LGTM, though I would prefer sharing sources between the two versions of `CompilerTraitAttribute.cs` and using an externally-defined constant to distinguish the two.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by TyOverby** [comment]
LGTM

File: src/Test/Utilities/Portable/CompilerTraitAttribute.cs (hunk 3 above)
Comment (gafter) [inline src/Test/Utilities/Portable/CompilerTraitAttribute.cs]: I was suggesting one use of the attribute and using a named constant for the assemblyName parameter. That named constant could have different values in the two assemblies.

**Comment by agocke** [comment]
:+1: Anything holding this up, aside from the merge conflict?

**Comment by jaredpar** [comment]
@agocke i just got behind and forgot to merge.  Rebasing now and will merge when tests pass.

**Comment by jaredpar** [comment]
ETA is failing due to a known build failure.  This change doesn't affect that code path (and ETA passed before my recent rebase).  Merging.

**Pull request merged**